File: Sources/WattsonCore/Quotas/Codebuff.swift
// Quotas/Codebuff.swift — Codebuff 订阅额度（移植自原生实现 CodebuffUsageFetcher）：
// 凭据 = env CODEBUFF_API_KEY 或 ~/.config/manicode/credentials.json（codebuff login 写入）；
// POST {base}/api/v1/usage（注意是 POST）+ 可选 GET {base}/api/user/subscription。
import Foundation

public struct CodebuffAuth: Sendable, Equatable {
    public var token: String
    public var baseUrl: String
    public init(token: String, baseUrl: String) {
        self.token = token
        self.baseUrl = baseUrl
    }
}

/// 凭据：env CODEBUFF_API_KEY 优先，其次 ~/.config/manicode/credentials.json
public func readCodebuffAuth(_ home: String, _ env: [String: String]) -> CodebuffAuth? {
    let envToken = env["CODEBUFF_API_KEY"]
    let baseUrl = (env["CODEBUFF_API_URL"]?.trimmingCharacters(in: .whitespaces).emptyToNil()
        ?? "https://www.codebuff.com").replacingOccurrences(of: "/+$", with: "", options: .regularExpression)
    if let envToken, !envToken.trimmingCharacters(in: .whitespaces).isEmpty {
        return CodebuffAuth(token: envToken.trimmingCharacters(in: .whitespaces), baseUrl: baseUrl)
    }
    guard let parsed = parseJSONFile((home as NSString).appendingPathComponent(".config/manicode/credentials.json")) else {
        return nil
    }
    let token = quotaStr(parsed["authToken"]) ?? quotaStr(parsed["default"]?["authToken"])
    return token.map { CodebuffAuth(token: $0, baseUrl: baseUrl) }
}

public func fetchCodebuffAccount(_ auth: CodebuffAuth, _ doFetch: FetchLike, _ now: Double) async -> QuotaAccount {
    let headers = [
        "authorization": "Bearer \(auth.token)",
        "accept": "application/json",
        "content-type": "application/json",
    ]
    do {
        let usage = try await quotaFetchJson(doFetch, "\(auth.baseUrl)/api/v1/usage", method: "POST",
                                             headers: headers,
                                             body: "{\"fingerprintId\":\"wattson-usage\"}")
        let u = usage.body
        let used = quotaNum(u["usage"]) ?? quotaNum(u["used"])
        let total = quotaNum(u["quota"]) ?? quotaNum(u["limit"])
        let remaining = quotaNum(u["remainingBalance"]) ?? quotaNum(u["remaining"])
        let nextQuotaReset = toEpochMs(u["next_quota_reset"])

        var windows: [QuotaWindow] = []
        if used != nil || total != nil || remaining != nil {
            let effectiveUsed: Double?
            if let used {
                effectiveUsed = used
            } else if let total, let remaining {
                effectiveUsed = total - remaining
            } else {
                effectiveUsed = nil
            }
            windows.append(QuotaWindow(
                key: "credits", label: "积分额度",
                total: total, used: used, remaining: remaining,
                percentage: percentOf(effectiveUsed, total),
                nextResetAt: nextQuotaReset))
        }

        // 订阅与周窗口是可选增强（原生实现给 2s grace），失败不拖垮主快照
        var planName: String?
        do {
            let sub = try await quotaFetchJson(doFetch, "\(auth.baseUrl)/api/user/subscription",
                                               headers: headers, timeoutMs: 5_000)
            let s = sub.body
            let subscription = s["subscription"]
            let rateLimit = s["rateLimit"]
            planName = quotaStr(subscription?["displayName"]) ?? quotaStr(s["displayName"])
                ?? quotaStr(subscription?["tier"]) ?? quotaStr(s["tier"])
            let weeklyUsed = quotaNum(rateLimit?["weeklyUsed"]) ?? quotaNum(rateLimit?["used"])
            let weeklyLimit = quotaNum(rateLimit?["weeklyLimit"]) ?? quotaNum(rateLimit?["limit"])
            if weeklyUsed != nil || weeklyLimit != nil {
                windows.append(QuotaWindow(
                    key: "week", label: "周额度",
                    total: weeklyLimit, used: weeklyUsed, remaining: nil,
                    percentage: percentOf(weeklyUsed, weeklyLimit),
                    nextResetAt: toEpochMs(rateLimit?["weeklyResetsAt"])))
            }
        } catch {
            // 订阅端点不可得时仅展示积分窗口
        }

        return finalizeAccount(kind: .codebuff, label: "Codebuff",
                               planName: planName, windows: windows, anySuccess: true, now: now)
    } catch {
        return finalizeAccount(kind: .codebuff, label: "Codebuff",
                               error: errText(error),
                               httpStatus: readHttpStatus(error).map(Double.init), now: now)
    }
}

public func codebuffAccount(home: String, env: [String: String], doFetch: FetchLike, now: Double) async -> QuotaAccount {
    guard let auth = readCodebuffAuth(home, env) else { return missingAccount(.codebuff, "Codebuff", now) }
    return await fetchCodebuffAccount(auth, doFetch, now)
}

File: server/src/quotas/codebuff.ts
// server/src/quotas/codebuff.ts — Codebuff 订阅额度（移植 CodexBar CodebuffUsageFetcher）：
// 凭据 = env CODEBUFF_API_KEY 或 ~/.config/manicode/credentials.json（codebuff login 写入）；
// POST {base}/api/v1/usage（注意是 POST）+ 可选 GET {base}/api/user/subscription。
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import {
  finalizeAccount, missingAccount, quotaFetchJson, readHttpStatus, errText,
  quotaNum, quotaStr, toEpochMs, percentOf,
  type FetchLike, type QuotaAccount, type QuotaWindow,
} from './types.js'

export type CodebuffAuth = { token: string; baseUrl: string }

export function readCodebuffAuth(home: string, env: NodeJS.ProcessEnv): CodebuffAuth | null {
  const envToken = env['CODEBUFF_API_KEY']
  const baseUrl = (env['CODEBUFF_API_URL']?.trim() || 'https://www.codebuff.com').replace(/\/+$/, '')
  if (envToken && envToken.trim()) return { token: envToken.trim(), baseUrl }
  let raw: string
  try {
    raw = readFileSync(join(home, '.config', 'manicode', 'credentials.json'), 'utf8')
  } catch {
    return null
  }
  let parsed: unknown
  try { parsed = JSON.parse(raw) } catch { return null }
  const root = parsed as Record<string, any>
  const token = quotaStr(root['authToken']) ?? quotaStr(root['default']?.['authToken'])
  return token ? { token, baseUrl } : null
}

export async function fetchCodebuffAccount(
  auth: CodebuffAuth, doFetch: FetchLike, now: number,
): Promise<QuotaAccount> {
  const headers: Record<string, string> = {
    authorization: `Bearer ${auth.token}`,
    accept: 'application/json',
    'content-type': 'application/json',
  }
  try {
    const usage = await quotaFetchJson(doFetch, `${auth.baseUrl}/api/v1/usage`, {
      method: 'POST', headers, body: JSON.stringify({ fingerprintId: 'wattson-usage' }),
    })
    const u = (usage.body ?? {}) as Record<string, any>
    const used = quotaNum(u['usage']) ?? quotaNum(u['used'])
    const total = quotaNum(u['quota']) ?? quotaNum(u['limit'])
    const remaining = quotaNum(u['remainingBalance']) ?? quotaNum(u['remaining'])
    const nextQuotaReset = toEpochMs(u['next_quota_reset'])

    const windows: QuotaWindow[] = []
    if (used !== null || total !== null || remaining !== null) {
      windows.push({
        key: 'credits', label: '积分额度',
        total, used, remaining,
        percentage: percentOf(used ?? (total !== null && remaining !== null ? total - remaining : null), total),
        nextResetAt: nextQuotaReset,
      })
    }

    // 订阅与周窗口是可选增强（CodexBar 给 2s grace），失败不拖垮主快照
    let planName: string | null = null
    try {
      const sub = await quotaFetchJson(doFetch, `${auth.baseUrl}/api/user/subscription`, { headers, timeoutMs: 5_000 })
      const s = (sub.body ?? {}) as Record<string, any>
      const subscription = (s['subscription'] ?? null) as Record<string, any> | null
      const rateLimit = (s['rateLimit'] ?? null) as Record<string, any> | null
      planName = quotaStr(subscription?.['displayName']) ?? quotaStr(s['displayName'])
        ?? quotaStr(subscription?.['tier']) ?? quotaStr(s['tier'])
      const weeklyUsed = quotaNum(rateLimit?.['weeklyUsed']) ?? quotaNum(rateLimit?.['used'])
      const weeklyLimit = quotaNum(rateLimit?.['weeklyLimit']) ?? quotaNum(rateLimit?.['limit'])
      if (weeklyUsed !== null || weeklyLimit !== null) {
        windows.push({
          key: 'week', label: '周额度',
          total: weeklyLimit, used: weeklyUsed, remaining: null,
          percentage: percentOf(weeklyUsed, weeklyLimit),
          nextResetAt: toEpochMs(rateLimit?.['weeklyResetsAt']),
        })
      }
    } catch {
      // 订阅端点不可得时仅展示积分窗口
    }

    return finalizeAccount({
      kind: 'codebuff', label: 'Codebuff', provider: null,
      planName, windows, anySuccess: true, now,
    })
  } catch (err) {
    return finalizeAccount({
      kind: 'codebuff', label: 'Codebuff', provider: null,
      error: errText(err), httpStatus: readHttpStatus(err), now,
    })
  }
}

export function codebuffAccount(home: string, env: NodeJS.ProcessEnv, doFetch: FetchLike, now: number): QuotaAccount | Promise<QuotaAccount> {
  const auth = readCodebuffAuth(home, env)
  if (!auth) return missingAccount('codebuff', 'Codebuff', now)
  return fetchCodebuffAccount(auth, doFetch, now)
}
